### ISSUE TITLE
CSV export of url report analysis

### ISSUE THREAD

**Issue by edjw** [openrightsgroup/cmp-issues#207]
As best we can, can we have CSV export functionality for the data that the report will be based on? @alexhaydock, if you can, could you list the pieces of data that you think supports the points you're making?  Then @dantheta, could you take that list and allow CSV export of that data

Feel free to suggest alternative ways of going about this btw

**Comment by alexhaydock** [comment]
@dantheta Already added a CSV export for the data held on the "ISP Report Stats" page as part of #200 

Are you thinking about raw data export?

**Comment by dantheta** [comment]
@edjw Is anything else required for this one?

**Comment by edjw** [comment]
It's more for @alexhaydock. Can we export CSVs to back up everything we say in the report?

**Comment by alexhaydock** [comment]
Not directly, I suppose. I will report back with the actual data ranges and values we've used to draw tables and graphs from and which ones would ideally be included in a CSV export so that we can do some very easy copy-pasting to ensure the figures are up-to-the-minute before the report goes public.

**Comment by alexhaydock** [comment]
Hi @dantheta 

I appreciate this might be a large ticket, but these are the data ranges we are using for graphs and tables in the document.

If you could produce a page that allows the export of all of these, that would be extremely useful - thanks!

I'm sorry this is a huge request, and that the data for the graphs is all contained in the tables that will be in the appendix, but for the purposes of handing the report to an external designer and for giving ourselves ranges of data to copy and paste to ensure that it's accurate right up to immediately before publication, including a table for each graph we've drawn in the final document seems like a good plan.

I've tried to express what I need here as cleanly as is possible but if you're getting stuck trying to work out what is needed (I wouldn't blame you), feel free to message me on RocketChat and I'll try and help out as soon as I can. Alternatively, I could forward you the work-in-progress document and you could take a look at the actual graphs we're using the data to draw if that would help you out.

---------

CSV export should include:

### For the appendix data
- The entirety of the "Reporter Summary" table from the [Report Stats](https://www.blocked.org.uk/control/ispreport/stats) page.
- The entirety of the "Damage Summary" table from the [Report Stats](https://www.blocked.org.uk/control/ispreport/stats) page.
- The entirety of the "Reports by ISP" table from the [Report Stats](https://www.blocked.org.uk/control/ispreport/stats) page.
- The entirety of the "ISP Reply Stats" table from the [Reply Stats](https://www.blocked.org.uk/control/ispreport/reply-stats) page.
- From the "Reply stats by network" table on the [Reply Stats](https://www.blocked.org.uk/control/ispreport/reply-stats) page, a secondary table should be produced using only the data for 2018, and which contains the following columns:
    - Network
    - Reports sent
    - Auto-replies logged
    - Total replies logged
    - Average reply interval
- From the "Reply stats by network" table on the [Reply Stats](https://www.blocked.org.uk/control/ispreport/reply-stats) page, a secondary table should be produced using only the data for 2018, and which contains the following columns:
    - Network
    - Unresolved reports
    - Unresolved non-policy reports

### For graphs
- A table using data from the "Reply stats by network" table on the [Reply Stats](https://www.blocked.org.uk/control/ispreport/reply-stats) page using only data from 2018, and with the following columns:
    - ISP name
    - Average reply interval (days)
- A table using data from the table from the [Reply Stats](https://www.blocked.org.uk/control/ispreport/reply-stats) page using only data from 2018, and with the following rows:
    - Reports unblocked
    - Reports rejected
    - Reports unresolved
- A table using data from the "Reply stats by network" table on the [Reply Stats](https://www.blocked.org.uk/control/ispreport/reply-stats) page using only data from 2018, and with the following columns:
    - ISP name
    - Reports Sent
    - Unresolved count
- A table using data from the "Reply stats by network" table on the [Reply Stats](https://www.blocked.org.uk/control/ispreport/reply-stats) page using only data from 2018, and with the following columns:
    - ISP name
    - Unresolved blocks which match policy
        - This will need to be calculated by subtracting the value in the "Unresolved non-policy block count" column from the data in the "Unresolved count" column.
    - Unresolved blocks which do not match policy
        - This is just the "Unresolved non-policy block count" column
- A table using data from the "Reporter Summary" table from the [Report Stats](https://www.blocked.org.uk/control/ispreport/stats) page for **2017**, and with the following rows:
    - Site Owner
        - This will need to be calculated as an aggregate of the "Site Owner (Business)", "Site Owner (Other)", and "Site Owner (Personal)" values.
    - Site User
    - Web Dev / Admin
    - Not Stated
        - This is currently recorded in that table as "Other/Unknown"
- A table which is the same as above, but for **2018**
- A table using data from the "Reporter Summary" table from the [Report Stats](https://www.blocked.org.uk/control/ispreport/stats) page for **all time**, and with the following rows:
    - Site Owner
        - Again, this will need to be calculated as an aggregate of the "Site Owner (Business)", "Site Owner (Other)", and "Site Owner (Personal)" values.
    - Site User
    - Web Dev / Admin
- A table using data from the "Reporter Summary" table from the [Report Stats](https://www.blocked.org.uk/control/ispreport/stats) page for **all time**, which plots the **total** reports submitted by users who are in one of the three "Site Owner" categories:
    - Business
    - Personal
    - Other
- And finally, a table which plots the data from the "Damage Summary" table from the [Report Stats](https://www.blocked.org.uk/control/ispreport/stats) page, using just the data from 2018. So it would have the following columns:
    - Damage type
    - Number of reports filed in 2018

**Comment by dantheta** [comment]
There's probably a few days of work here.  Are there any sections which can be highlighted as priority for delivery first?

**Comment by alexhaydock** [comment]
I think the part that I've headed "appendix data" would be the most useful first. I can extrapolate the rest of the data from there.

**Comment by JimKillock** [comment]
Could I suggest, if this is a lot of work, that we try to avoid doing it, as we have a delivery deadline of Weds? 

Is there a workaround for instance? Even if that is cut and paste …

**Comment by alexhaydock** [comment]
> Is there a workaround for instance? Even if that is cut and paste …

One potential workaround would be to format the tables on the Blocked backend so that they are copy-pastable into Excel / Google Sheets etc.

Currently copy/paste operations on those tables result in a single long string of text being copied.

**Comment by JimKillock** [comment]
Ok, yes. Another workaround would be for Dan to manually export the relevant data at a particular time.

**Comment by dantheta** [comment]
Using Ctrl+Click (cmd+click on mac) on a table in firefox allows a table to be copied to the clipboard in a spreadsheet compatible format (I actually started typing that before Alex's comment :) )

**Comment by alexhaydock** [comment]
> Using Ctrl+Click (cmd+click on mac) on a table in firefox allows a table to be copied to the clipboard in a spreadsheet compatible format (I actually started typing that before Alex's comment :) )

This is very useful to know! In that case, I can probably compile the data myself reasonably quickly from the tables we already have available in the backend. I can work with what we have and compile the data for the designer and to make sure our report is up-to-date right before publication.

In that case, we can discard the above, and convert this ticket to some other housekeeping issues that will help me get the right data ranges:

- Add a column to the "Reply stats by network" with the title "Unresolved policy block count" (i.e. Unresolved blocks which match policy (or at least have not been marked as not matching it)
- Add a row to the "Reporter Summary" table which totals up the members of the three "Site Owner" subcategories (Business, Other, Personal)
- And finally something which doesn't make sense to open a separate ticket for but are you able to rename the "Other/Unknown" reporter category to "Not Stated"? That sounds a bit better.

Thanks :+1:

**Comment by dantheta** [comment]
Reporter category rename done.

**Comment by dantheta** [comment]
unresolved policy block count added.

**Comment by dantheta** [comment]
The extra site owner total row is now present,
